### PR TITLE
Allow for global objects with a mutable prototype chain

### PR DESCRIPTION
**Pull request by Ms2ger**
This is required to define the ShadowRealm global object in IDL.

- [ ] At least two implementers are interested (and none opposed):
   * …
   * …
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * https://github.com/web-platform-tests/wpt/pull/43756
- [ ] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chromium: …
   * Gecko: …
   * WebKit: …
   * Deno: …
   * Node.js: …
   * webidl2.js: …
   * widlparser: …
- [ ] [MDN issue](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) is filed: …
- [ ] The top of this comment includes a [clear commit message](https://github.com/whatwg/meta/blob/main/COMMITTING.md) to use. 

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)

***
<a href="https://whatpr.org/webidl/1381.html" title="Last updated on Jan 11, 2024, 4:32 PM UTC (56966c3)">Preview</a> | <a href="https://whatpr.org/webidl/1381/8e9affa...56966c3.html" title="Last updated on Jan 11, 2024, 4:32 PM UTC (56966c3)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -9687,6 +9687,10 @@ from other interfaces. Specifically:
     will correspond to properties on the object itself rather than on
     [=interface prototype objects=].
 
+A [=realm=] can be defined to
+<dfn export for="realm" lt="has a mutable global prototype chain">have a mutable global prototype chain</dfn>.
+If a [=realm=] is not defined as such, the prototype of its global object can not be changed.
```
Comment (annevk) [inline index.bs:9692]: I keep having to learn that conditional fields I have added in the past have led to some kind of issue that required changing them to always-present fields. I would suggest we make this an always-present field of realms with a default of false.

> Each realm has an associated "uses mutable global prototype chain" (a boolean). It is initially false.
>
> Note: This allows the ShadowRealm global to have a mutable prototype.

File: index.bs [hunk 2]
```diff
@@ -9687,6 +9687,11 @@ from other interfaces. Specifically:
     will correspond to properties on the object itself rather than on
     [=interface prototype objects=].
 
+All [=realms=] have a <dfn export for="realm">is global prototype chain mutable</dfn> flag,
```
Comment (annevk) [inline index.bs:9690]:
```suggestion
All [=realms=] have an <dfn export for="realm">is global prototype chain mutable</dfn> boolean,
```
No more flags please!

**Review by annevk — COMMENTED** [review]
Thanks, pretty close now. (I'm trusting the JS bits as you probably know that better than me.)

**Review by annevk — APPROVED** [review]
Let's give some time for others to comment, but to me this seems good to go.

**Comment by domenic** [comment]
> I guess there's a question of how much we want to wait for implementer interest for an infrastructure change. @domenic thoughts?

In general I'm supportive of not waiting too much. (E.g. it seems annoying to me that #1353 is stalled on that.) But the overall fate of ShadowRealm seems a bit unclear to me, given its regression to stage 2, so I'm a bit hesitant about making a bunch of ShadowRealm-related changes to our specs.

But, it's probably fine, as long as we're good about reverting if a proposal gets withdrawn. That's what we did for, e.g., https://github.com/whatwg/webidl/pull/924.

**Comment by Ruben-VandeVelde** [comment]
ShadowRealms was reverted to stage 2 because the web-side specifications were not ready; this PR is one of the blockers to advance to stage 3 again.

**Comment by Ms2ger** [comment]
ShadowRealms was reverted to stage 2 because the web-side specifications were not ready; this PR is one of the blockers to advance to stage 3 again.

**Comment by annevk** [comment]
@Ms2ger would you be willing to file the implementation bugs for this and update OP afterwards?

**Pull request merged**